website/blog/2025/04-17-Leaner-Clusters-Lower-Bills.md

```diff
@@ -0,0 +1,315 @@
+
+### Consolidation: Avoiding a Fragmented Seed Landscape
+
+If possible, avoid operating too many small seeds unless required by regulations or driven by end-user demand. As Gardener supports control plane migration, you can consolidate your control planes into fewer, larger seeds where reasonable. Since starting Gardener in production in 2017, we've encountered technological advancements (e.g., Azure Availability Sets to Zones) and corrected initial misconfigurations (e.g., too-small CIDR ranges limiting pod/node counts) that necessitated recreating seeds. While hard conflicts (like seed/shoot cluster IP address overlaps) can sometimes block migration to differently configured seeds, you can often at least merge multiple seeds into one or fewer. The key takeaway is that a less fragmented seed landscape generally leads to better efficiency.
+
+However, there is a critical caveat: Gardener allows control planes to reside in different regions (or even different cloud providers) than their worker nodes. This flexibility comes at the cost of inter-regional or internet network traffic. These additional network-related costs can easily negate efficiency gains from seed consolidation. Therefore, consolidate thoughtfully, being mindful that excessive consolidation across regions can significantly increase network costs (intra-region traffic is cheaper than inter-region traffic, and internet traffic is usually the most expensive).
+
+## Quick Wins in Networking and Storage
+
+While compute was our main focus, we also addressed significant cost drivers in networking and storage early on.
+
+### Centralized Ingress & Caching
+
+*   **Centralized Ingress:** In Gardener's early days, each shoot control plane had its own Load Balancer (LB), plus another for the reverse tunnel connection to worker nodes (to reach webhooks, scrape metrics, stream logs, `exec` into pods, etc.). This proliferation of LBs was expensive. We transitioned to a model using a central Istio ingress-gateway per seed cluster with a single LB, leveraging SNI (Server Name Indication) routing to direct traffic to the correct control plane API servers. We also reversed the connection direction: shoots now connect *to* seed clusters, and seeds connect *to* the garden cluster. This reduced the need for LBs exposing seed components and enabled *private* shoots or even *private* seeds behind firewalls.
+*   **Registry Cache:** Pulling container images for essential components (like CNI, CSI drivers, kube-proxy) on every new node startup generated significant network traffic and costs. We implemented a [registry cache extension](https://github.com/gardener/gardener-extension-registry-cache), drastically reducing external image pulls (see [Issue #3](https://github.com/gardener/gardener-extension-registry-cache/issues/3)).
```

Comment:
You may want to reference the Registry Cache blog post as well: https://gardener.cloud/blog/2024/04-22-gardeners-registry-cache-extension-another-cost-saving-win-and-more/

Reply:
Absolutely, much better idea. I replaced the issue link with the blog post.